cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBPartitionManager.java

```diff
@@ -0,0 +1,291 @@
+import org.iq80.leveldb.DB;
+import org.iq80.leveldb.Options;
+import org.iq80.leveldb.impl.Iq80DBFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * Manages partitions of a logical MessageTable.
+ * Each partition is a physical LevelDB table in a directory structure like:
+ *
+ *   [base dir]/[namespace].[tablename].[topic].[generation]/part.[start].[end]
+ *
+ * Each partition contains messages with a publish time between the start (inclusive) and end (exclusive) timestamps.
+ */
+public class LevelDBPartitionManager implements Closeable {
+  private static final Iq80DBFactory LEVEL_DB_FACTORY = Iq80DBFactory.factory;
+  private static final String PART_PREFIX = "part.";
+  private final File topicDir;
+  private final Options dbOptions;
+  private final long partitionSizeMillis;
+  private final ConcurrentNavigableMap<Long, LevelDBPartition> partitions;
+  private final AtomicBoolean initialized;
+
+  public LevelDBPartitionManager(File topicDir, Options dbOptions, long partitionSizeMillis) {
+    this.topicDir = topicDir;
+    this.dbOptions = dbOptions;
+    this.partitionSizeMillis = partitionSizeMillis;
+    this.partitions = new ConcurrentSkipListMap<>();
+    this.initialized = new AtomicBoolean(false);
+  }
+
+  /**
+   * Prune partitions that have an end timestamp older than or equal to the given threshold timestamp.
+   */
+  public int prunePartitions(long thresholdTimestamp) throws IOException {
+    File[] partitionDirs = topicDir.listFiles();
+    int numPruned = 0;
+    if (partitionDirs == null) {
+      return numPruned;
+    }
+    for (File partitionDir : partitionDirs) {
+      String partitionName = partitionDir.getName();
+      TimeInterval timeInterval = extractIntervalFromName(partitionName);
+      if (timeInterval == null) {
+        // can only happen if somebody manually created some files/directories here
+        continue;
+      }
+      if (thresholdTimestamp >= timeInterval.endMillis) {
+        DirUtils.deleteDirectoryContents(partitionDir);
+        partitions.remove(timeInterval.startMillis);
+        numPruned++;
+      }
+    }
+    return numPruned;
+  }
+
+  public LevelDBPartition getOrCreatePartition(long publishTime) throws IOException {
+    Long partitionKey = partitions.floorKey(publishTime);
+    if (partitionKey != null) {
+      LevelDBPartition partition = partitions.get(partitionKey);
+      if (partition.getEndTime() > publishTime) {
+        return partition;
+      }
+    }
+
+    synchronized (this) {
+      // Check again to make sure no new instance was being created while this thread is acquiring the lock
+      partitionKey = partitions.floorKey(publishTime);
```

Comment:
Since this code is very similar to the previous block, maybe it would be useful to move this into a separate function.